docs/advanced/runner.md

```diff
@@ -0,0 +1,148 @@
+  /**
+   * Called after collecting tests and before "onBeforeRun".
+   */
+  onCollected?(files: File[]): unknown
+
+  /**
+   * Called before running a single test. Doesn't have "result" yet.
+   */
+  onBeforeRunTest?(test: Test): unknown
+  /**
+   * Called before actually running the test function. Already has "result" with "state" and "startTime".
+   */
+  onBeforeTryTest?(test: Test, retryCount: number): unknown
+  /**
+   * Called after result and state are set.
+   */
+  onAfterRunTest?(test: Test): unknown
+  /**
+   * Called right after running the test function. Doesn't have new state yet. Will not be called, if the test function throws.
+   */
+  onAfterTryTest?(test: Test, retryCount: number): unknown
+
+  /**
+   * Called before running a single suite. Doesn't have "result" yet.
+   */
+  onBeforeRunSuite?(suite: Suite): unknown
+  /**
+   * Called after running a single suite. Has state and result.
+   */
+  onAfterRunSuite?(suite: Suite): unknown
+
+  /**
+   * If defined, will be called instead of usual Vitest suite partition and handling.
+   * "before" and "after" hooks will not be ignored.
+   */
+  runSuite?(suite: Suite): Promise<void>
+  /**
+   * If defined, will be called instead of usual Vitest handling. Useful, if you have your custom test function.
+   * "before" and "after" hooks will not be ignored.
+   */
+  runTest?(test: Test): Promise<void>
+
+  /**
+   * Called, when a task is updated. The same as "onTaskUpdate" in a reporter, but this is running in the same thread as tests.
+   */
+  onTaskUpdate?(task: [string, TaskResult | undefined][]): Promise<void>
+
+  /**
+   * Called before running all tests in collected paths.
+   */
+  onBeforeRun?(files: File[]): unknown
+  /**
+   * Called right after running all tests in collected paths.
+   */
+  onAfterRun?(files: File[]): unknown
+  /**
+   * Called when new context for a test is defined. Useful, if you want to add custom properties to the context.
+   * If you only want to define custom context with a runner, consider using "beforeAll" in "setupFiles" instead.
+   */
+  extendTestContext?(context: TestContext): TestContext
+  /**
+   * Called, when files are imported. Can be called in two situations: when collecting tests and when importing setup files.
+   */
+  importFile(filepath: string, source: VitestRunnerImportSource): unknown
+  /**
+   * Publically available configuration.
+   */
+  config: VitestRunnerConfig
+}
+```
+
+When initiating this class, Vitest passes down Vitest config, - you should expose it as a `config` property.
+
+::: warning
+`importFile` method in your custom runner must be inlined in `deps.inline` config option, if you call Node `import` inside.
+:::
+
+::: tip
+Snapshot support and some other features depend on the runner. If you don't want to lose it, you can extend your runner from `VitestTestRunner` imported from `vitest/runners`. It also exposes `BenchmarkNodeRunner`, if you want to extend benchmark functionality.
```

Comment:
Is there a more concrete example of this?

```js
import { VitestRunner } from '@vitest/runner';
         ^^^^^^^^^^^^
SyntaxError: The requested module '@vitest/runner' does not provide an export named 'VitestRunner'
```

I don't see VitestRunner exported in https://unpkg.com/@vitest/runner@0.28.4/dist/index.js